packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts

```diff
@@ -38,13 +37,114 @@ export class ObjectForest extends SimpleDependee implements IEditableForest {
     }
 
     applyDelta(delta: Delta.Root): void {
-        throw new Error("Method not implemented.");
+        // TODO: refactor object forest to use root node above detached fields, like how PathNode works.
+        // Then factor out this editing code to work on any such JsonableTree.
+
+        this.beforeChange();
+        const moves: Map<Delta.MoveId, DetachedField> = new Map();
+        const currentNode: Cursor = this.allocateCursor();
+        let currentField: FieldKey | undefined;
+        const moveIn = (index: number, toAttach: DetachedField): number => {
+            assert(currentField !== undefined, "must be in field to onMoveIn");
+            const children = this.roots.get(toAttach) ?? fail("Can not attach non-existent range");
+            this.roots.delete(toAttach);
+            let dstField: ObjectField;
+            if (currentNode.state === ITreeSubscriptionCursorState.Cleared) {
+                const dst = currentField as unknown as DetachedField;
+                assert(toAttach !== dst, "can not attach range to itself");
+                // TODO: protect against parenting under itself creating a cycle as well:
+                // or maybe that should delete the subtree?
+                dstField = this.getRoot(dst);
+            } else {
+                if (children.length === 0) {
+                    return 0; // Prevent creating 0 sized fields when inserting empty into empty.
+                }
+                dstField = getGenericTreeField(currentNode, currentField, true);
+            }
+            assertValidIndex(index, dstField, true);
+            // TODO: this will fail for very large moves due to argument limits.
+            dstField.splice(index, 0, ...children);
+
+            return children.length;
+        };
+        const visitor = {
+            onDelete: (index: number, count: number): void => {
+                assert(currentField !== undefined, "must be in field to onDelete");
+                visitor.onMoveOut(index, count);
+            },
+            onInsert: (index: number, content: Delta.ProtoNode[]): void => {
+                assert(currentField !== undefined, "must be in field to onInsert");
+                const range = this.add(content.map((data) => new TextCursor(data)));
+                moveIn(index, range);
+            },
+            onMoveOut: (index: number, count: number, id?: Delta.MoveId): void => {
+                assert(currentField !== undefined, "must be in field to onMoveOut");
+                let srcField: ObjectField;
+                if (currentNode.state === ITreeSubscriptionCursorState.Cleared) {
+                    srcField = this.getRoot(currentField as unknown as DetachedField);
+                } else {
+                    srcField = getGenericTreeField(currentNode, currentField, false);
+                }
+                const field = this.detachRangeOfChildren(srcField, index, index + count);
```

Comment:
How do we ensure that the DetachedField created here doesn't collide with a move destination referenced in the delta?

Reply:
they are separate scopes. In here we never use move ids as detached fields or field keys, we instead allocate a new unique detached field for each move id and store them in the `moves` map.

Reply:
On line 60 we interpret a FieldKey specified by the delta as a DetachedField, and then edit that field. Even if we assume that there were no detached fields at the start of applying the delta, isn't there a risk that the specified detached field was previously created to be used to temporarily store moved out content?